### PR TITLE
Event Streaming

### PR DESCRIPTION
**Pull request by blinky3713**
The goal of this PR is to propose some changes to allow for more flexible event monitoring/streaming. I'm hoping that @XertroV @kejace @iostat @cmditch can take a look, comment on the design, and point out any flaws that might be here.

## Changes

There are a number of breaking changes.

1. create a newtype for `BlockNumber` around `BigNumber`

2. change the `Filter` object to require the `toBlock` and `fromBlock`, rather than `Maybe`. The fact that you _can_ get away with using the geth defaults if the filter fields are not specified is not a good reason to be lazy, and it made dealing with events in the rest of these changes way harder and more ad-hoc.

3. rename `CallMode` to `BlockMode` and add an `Ord` instance.

4. Use the [purescript-machines](https://github.com/purescript-contrib/purescript-machines) library to standardize the way we are streaming events, rather than the ad-hoc polling. The goal is to allow for seamlessly streaming events from an arbitrary `BlockMode` to an arbitrary 'BlockMode'. Take the following examples. Say we would like to stream 'Transfer' events from an ERC20 contract, and we are currently at block 15.

- Stream events from block 10 to block 20. The event streamer should use `eth_getLogs` to fetch the first blocks 10 ... 15, then create a new filter and use `eth_getFilterChanges` for blocks 15 ... 20 and perform the appropriate clean up.

- We would like to stream events from block 0 to block 10. We should use `eth_getLogs` for everything.

- We would only like to stream `Latest` to `Latest`, the streamer should make a new filter, then call `eth_getFilterChanges` for everything.

- We would like to stream 18 to `Latest`. It should make a new filter and call `eth_getFilterChanges` for everything.

Furthermore, we would like the option of specifying a range of blocks for anything which uses `eth_getLogs` rather than going block by block. You should be able to derive the normal case by specifying the range to be 1.

### CODE REVIEW

File: src/Network/Ethereum/Web3/Contract.purs [hunk 1]
```diff
@@ -46,34 +57,156 @@ class EventFilter a where
...
+      if s.currentBlock > end
+         then pure Halt
+         else let to' = newTo end s.currentBlock s.windowSize
+                  fltr = s.initialFilter
+                           # _fromBlock .~ BN s.currentBlock
+                           # _toBlock .~ BN to'
+              in pure $ Emit fltr $ mealy \s' ->
+                   filterStream' s' {currentBlock = succ to'}
+
+-- | Coerce a 'BlockMode' to an actual 'BlockNumber'.
```
Comment (kejace) [inline src/Network/Ethereum/Web3/Contract.purs]: I think we want to export this function as well

**Comment by kejace** [comment]
I just modified a branch of `token.monitor` to use the latest code.

I have an eventFilter

```purescript
let fltTransfer = eventFilter (Proxy :: Proxy ERC20.Transfer) address
                    # _fromBlock .~ Latest
                    # _toBlock .~ Latest
```
and I'm seeing events as per previous behavior 👍.

I'd be happy to take on modifying the generator at this stage to reflect the changes.
I also think we should actually merge the generator into this package by now.

**Comment by kejace** [comment]
I think it's worth considering how the `par` branch would interplay with this new streaming branch, and tests we could write for the two together. Happy to merge, but will let others comment before doing so.

**Comment by kejace** [comment]
Cases to confirm  (for current block `m` and `n = m - 20`) assuming fully synched node.

- [x]  `[n, n+10]` 
- [x] `[n, Latest)`
- [ ] `[m+5,m+10]` seems to give blocks for `m + 1` for example
- [ ] `[m+5,Latest)`

**Comment by blinky3713** [comment]
@kejace with respect to the par branch, i think it wont be hard to rebase those changes on this, but this should settle first

File: src/Network/Ethereum/Web3/Types/Types.purs [hunk 2]
```diff
@@ -282,7 +319,6 @@ newtype TransactionOptions =
                      }
 
 derive instance genericTransactionOptions :: Generic TransactionOptions _
-derive instance newtypeTransactionOptions :: Newtype TransactionOptions _
```
Comment (XertroV) [inline src/Network/Ethereum/Web3/Types/Types.purs]: Is this intentionally removed? Is the idea to use lenses in general? (Note: I like lenses but without template purescript or some automatic method of derivation they're a bit annoying)

**Comment by XertroV** [comment]
I don't use filters atm so I don't have too many comments on exactly how they're exposed / used.

1. `BlockNumber`: This seems sensible. I think it's good.

2. Yup - explicitness is good!

3. Note sure `BlockMode` is the best name for it, maybe something like `ChainLocation`?

4. I've not used purescript-machines so can't comment on that choice, but I think standardising is the right way to go. The cases you lay out seem very sensible.

Also 👍 re using ranges and minimising calls. Doesn't matter much for local nodes but if you're using infura or something then it's much more elegant (and easier to remain consistent in the case of a chain reorg, I suspect).

All in all looks good!

File: src/Network/Ethereum/Web3/Types/Types.purs (hunk 2 above)
Comment (blinky3713) [reply src/Network/Ethereum/Web3/Types/Types.purs]: no this is not intentional

**Comment by blinky3713** [comment]
ok so it seems like we can start to work on the generator and tests with the design as is, then iterate once the tests are up

It seems like the main concerns outside of testing are
 - avoid excessive `Newtype` instances, or at least expose all the constructors when it makes sense. 
- Possibly change the name of 'BlockMode'. Proposed name was 'ChainLocation', i sort of like 'ChainIndex'. 

@kejace has started on working on some of the tests, but we need to distribute the work on tests and generator.

**Comment by XertroV** [comment]
@blinky3713 I like `ChainIndex` 

Also, how is the generator affected by this PR? (Or do you mean integrating it with this repo?)

**Comment by blinky3713** [comment]
@XertoV the generator is not _really_ affected, just fixing how the default instances are generated for the `EventFilter` class

**Comment by kejace** [comment]
We updated `web3-generator` and `web3-tests` and everything passes. Will merge.

**Pull request merged**